File: Team Attendance Management System Requirements (1)/attendance.py
from flask import Blueprint, render_template, redirect, url_for, request, flash, jsonify
from flask_login import login_required, current_user
from app.models.models import Attendance, db
from datetime import datetime
import json
from geopy.geocoders import Nominatim

attendance = Blueprint('attendance', __name__)

@attendance.route('/')
@login_required
def index():
    return render_template('index.html')

@attendance.route('/check-in', methods=['POST'])
@login_required
def check_in():
    latitude = request.form.get('latitude')
    longitude = request.form.get('longitude')
    notes = request.form.get('notes', '')
    
    if not latitude or not longitude:
        flash('Location data is required')
        return redirect(url_for('attendance.index'))
    
    # Check if user already has an open attendance record
    open_attendance = Attendance.query.filter_by(
        user_id=current_user.id, 
        check_out_time=None
    ).first()
    
    if open_attendance:
        flash('You already have an active check-in')
        return redirect(url_for('attendance.index'))
    
    # Try to get address from coordinates
    address = None
    try:
        geolocator = Nominatim(user_agent="attendance_system")
        location = geolocator.reverse(f"{latitude}, {longitude}")
        if location:
            address = location.address
    except Exception as e:
        # Just log the error, but continue with check-in
        print(f"Error getting address: {e}")
    
    # Create new attendance record
    new_attendance = Attendance(
        user_id=current_user.id,
        check_in_time=datetime.utcnow(),
        notes=notes,
        location_address=address
    )
    new_attendance.set_location(latitude, longitude)
    
    db.session.add(new_attendance)
    db.session.commit()
    
    flash('Check-in successful')
    return redirect(url_for('attendance.index'))

@attendance.route('/check-out', methods=['POST'])
@login_required
def check_out():
    # Find the user's open attendance record
    open_attendance = Attendance.query.filter_by(
        user_id=current_user.id, 
        check_out_time=None
    ).first()
    
    if not open_attendance:
        flash('No active check-in found')
        return redirect(url_for('attendance.index'))
    
    # Update the check-out time
    open_attendance.check_out_time = datetime.utcnow()
    db.session.commit()
    
    flash('Check-out successful')
    return redirect(url_for('attendance.index'))

@attendance.route('/history')
@login_required
def history():
    # Get user's attendance records
    records = Attendance.query.filter_by(user_id=current_user.id).order_by(Attendance.check_in_time.desc()).all()
    return render_template('history.html', records=records)

@attendance.route('/api/attendance/status')
@login_required
def status():
    # Check if user has an open attendance record
    open_attendance = Attendance.query.filter_by(
        user_id=current_user.id, 
        check_out_time=None
    ).first()
    
    if open_attendance:
        return jsonify({
            'status': 'checked_in',
            'check_in_time': open_attendance.check_in_time.isoformat(),
            'location': open_attendance.get_location(),
            'location_address': open_attendance.location_address
        })
    
    return jsonify({'status': 'checked_out'})

File: Team Attendance Management System Requirements (1)/attendance.ts
import { getCloudflareContext } from '@/lib/cloudflare';

// Attendance type definition
export interface Attendance {
  id: number;
  user_id: number;
  check_in_time: string;
  check_out_time: string | null;
  location_data: string;
  location_address: string | null;
  notes: string | null;
}

// Location type definition
export interface Location {
  latitude: number;
  longitude: number;
}

// Check in function
export async function checkIn(userId: number, location: Location, notes?: string): Promise<Attendance | null> {
  const { env } = getCloudflareContext();
  
  // Check if user already has an open attendance record
  const openAttendance = await env.DB.prepare(
    'SELECT id FROM attendance WHERE user_id = ? AND check_out_time IS NULL'
  ).bind(userId).first();
  
  if (openAttendance) {
    return null; // Already checked in
  }
  
  // Store location as JSON string
  const locationData = JSON.stringify(location);
  
  // Insert new attendance record
  const result = await env.DB.prepare(
    'INSERT INTO attendance (user_id, location_data, notes) VALUES (?, ?, ?)'
  ).bind(userId, locationData, notes || null).run();
  
  if (!result.success) {
    return null;
  }
  
  // Get the newly created attendance record
  const newAttendance = await env.DB.prepare(
    'SELECT id, user_id, check_in_time, check_out_time, location_data, location_address, notes FROM attendance WHERE id = ?'
  ).bind(result.meta.last_row_id).first();
  
  return newAttendance as Attendance;
}

// Check out function
export async function checkOut(userId: number): Promise<Attendance | null> {
  const { env } = getCloudflareContext();
  
  // Find the user's open attendance record
  const openAttendance = await env.DB.prepare(
    'SELECT id FROM attendance WHERE user_id = ? AND check_out_time IS NULL'
  ).bind(userId).first();
  
  if (!openAttendance) {
    return null; // No active check-in
  }
  
  // Update the check-out time
  const result = await env.DB.prepare(
    'UPDATE attendance SET check_out_time = CURRENT_TIMESTAMP WHERE id = ?'
  ).bind(openAttendance.id).run();
  
  if (!result.success) {
    return null;
  }
  
  // Get the updated attendance record
  const updatedAttendance = await env.DB.prepare(
    'SELECT id, user_id, check_in_time, check_out_time, location_data, location_address, notes FROM attendance WHERE id = ?'
  ).bind(openAttendance.id).first();
  
  return updatedAttendance as Attendance;
}

// Get attendance status
export async function getAttendanceStatus(userId: number): Promise<{ status: 'checked_in' | 'checked_out', attendance?: Attendance }> {
  const { env } = getCloudflareContext();
  
  // Find the user's open attendance record
  const openAttendance = await env.DB.prepare(
    'SELECT id, user_id, check_in_time, check_out_time, location_data, location_address, notes FROM attendance WHERE user_id = ? AND check_out_time IS NULL'
  ).bind(userId).first();
  
  if (!openAttendance) {
    return { status: 'checked_out' };
  }
  
  return { 
    status: 'checked_in',
    attendance: openAttendance as Attendance
  };
}

// Get attendance history
export async function getAttendanceHistory(userId: number, limit = 50): Promise<Attendance[]> {
  const { env } = getCloudflareContext();
  
  // Get user's attendance records
  const records = await env.DB.prepare(
    'SELECT id, user_id, check_in_time, check_out_time, location_data, location_address, notes FROM attendance WHERE user_id = ? ORDER BY check_in_time DESC LIMIT ?'
  ).bind(userId, limit).all();
  
  return records.results as Attendance[];
}

// Calculate duration in hours
export function calculateDuration(checkInTime: string, checkOutTime: string | null): number | null {
  if (!checkOutTime) {
    return null;
  }
  
  const checkIn = new Date(checkInTime);
  const checkOut = new Date(checkOutTime);
  const durationMs = checkOut.getTime() - checkIn.getTime();
  
  return parseFloat((durationMs / (1000 * 60 * 60)).toFixed(2));
}

// Parse location data
export function parseLocationData(locationData: string): Location | null {
  try {
    return JSON.parse(locationData) as Location;
  } catch (error) {
    return null;
  }
}
